### PR TITLE
Update YesSQL to 3.2.0

### PR DESCRIPTION
**Pull request by hishamco**
https://github.com/sebastienros/yessql/releases/tag/v3.1.0

### CODE REVIEW

**Comment by hishamco** [comment]
Seems there's a breaking changes

**Comment by hishamco** [comment]
There's a `schema` parameter has been added

**Comment by sebastienros** [comment]
I might do another update to fix the breaking changes, defaulting to the previous values, and also setting `UseLegacyIdentityColumn` to true by default. Then only do these breaking changes in a major version.

**Comment by sebastienros** [comment]
Also please don't merge without my approval.

**Comment by hishamco** [comment]
Sure, I will not merge until we fix the breaking chnages or using the default schema, also after your approval

**Comment by sebastienros** [comment]
Please use Configuration.Schema when the schema is requested (or null)

**Comment by hishamco** [comment]
Is there a new update released?

**Comment by sebastienros** [comment]
I just pushed 3.2.0, should appear soon.

**Comment by hishamco** [comment]
I was totally busy last few days, but I need to fix it soon

**Comment by sebastienros** [comment]
@MikeAlhayek FYI since you have commit rights now, you can pull Hisham's branch and update it if needed. You can even change the files directly from GH.

**Comment by MikeAlhayek** [comment]
@sebastienros I pinged @hishamco privately. He should have this PR updated with in an hour

**Comment by hishamco** [comment]
Seb, I will update the PR ASAP, since I'm working on other testing API. BTW GH allowed editing the changed files AFAIK, but I need to change other files

**Comment by hishamco** [comment]
Seb, I will update the PR ASAP, since I'm working on other testing API. BTW GH allowed editing the changed files AFAIK, but I need to change other files

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:151]: Remove ` IOptions<YesSqlOptions> yesSqlOptions,` from being injected and use the following instead

```suggestion
        selectBuilder.Table(_yesSqlConfiguration.TableNameConvention.GetDocumentTable(), alias: null, _yesSqlConfiguration.Schema);
```

File: src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs:169]: I would suggest you use the same enum used by YesSql `IdentityColumnSize` instead of using a boolean `UseLegacyIdentityColumn`  maybe at somepoint YesSql will support other enum values like string.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs:169]: It's column size not type, I think it's fine to stay with a boolean value, but I never mind to use the `IdentityColumnSize `

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:151]: I think we still need `YesSqlOptions` for table prefix separator

**Comment by hishamco** [comment]
@jtkech I expecting `YesSql.IConfiguration` is registered by default, am I right or shall I add the registeration myself to make the tests pass?

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Migrations.cs:180]: the schema can be accessed through `session.Store.Configuration.Schema` the same way `tablePrefix` is used a few lines above.

File: src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Predicates/PredicateQuery.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/Predicates/PredicateQuery.cs:130]: If `tablePrefix` is not used here, it's probably because this is just an alias, and then schema can/should be null.

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:151]: yeah true. but I would read the `TableNameConvention` from the `IConfiguration` instead

File: src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs:23]: I think that where @MikeAlhayek 's suggestion might apply. Store an enum, or the Type of the identity column. Today it is int32/int63 but maybe we'll support Guid someday.

File: src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs:169]: yes but it is an enum value which can have more than 2 values. Maybe it does not have the best name in YesSql but it should also be an enum in OC not bool.

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: Weird to need both "YesSqlOptions" and "YeslSqlConfiguration"

File: src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs:23]: Exactly. We should not use bool in OC and should use the same Enum used in YesSql.

File: test/OrchardCore.Tests/Orchard.Queries/SqlParserTests.cs
Comment (sebastienros) [inline test/OrchardCore.Tests/Orchard.Queries/SqlParserTests.cs:35]: Pass the schema directly, this is what is done for dialect and prefix.

Comment (sebastienros) [reply test/OrchardCore.Tests/Orchard.Queries/SqlParserTests.cs:35]: And then the test can verify that the schema is used in the result, like we check the prefix.

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: `YesSqlOptions` has the option [TablePrefixSeparator ](https://github.com/OrchardCMS/OrchardCore/blob/8d9ced24fe9c815ccbc309cdf2a48db37db6a1aa/src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs#L21) which isn't used in `YesSql`. This is the reason both are needed. I think we should use `IConfiguration` to get the currently configured services instead of using `YesSqlOptions` since `YesSqlOptions` could have null configurations like a null value for `TableNameConvention`

**Comment by MikeAlhayek** [comment]
> @jtkech I expecting `YesSql.IConfiguration` is registered by default, am I right or shall I add the registeration myself to make the tests pass?

You can inject `ISession` and assess it `_session.Store.Configuration` instead of injecting `YesSql.IConfiguration` directly

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: I may be wrong but `YesSqlOptions` provides options to configure the yesSql store configuration.

So why not just usin the `IStore` to retrieve what is actually used.

From memory, I did a PR to ensure that `TableNameConvention` is in sync

                // Ensure a non null 'TableNameConvention' to be always used for `YesSql.Configuration` and then in sync with it.
                services.PostConfigure<YesSqlOptions>(o => o.TableNameConvention ??= new YesSql.Configuration().TableNameConvention);

For `TablePrefixSeparator` it has a default value, and it is used indirectly by YesSql through the TablePrefix.

                    if (!String.IsNullOrWhiteSpace(shellSettings["TablePrefix"]))
                    {
                        var tablePrefix = shellSettings["TablePrefix"].Trim() + yesSqlOptions.TablePrefixSeparator;
                        storeConfiguration = storeConfiguration.SetTablePrefix(tablePrefix);
                    }

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: Yes `TableNameConvention` should be retrieved from the `ISession.Store.Configuration` not from `YesSqlOptions`. But we should still use `YesSqlOptions` to determine the prefix-separator that the tenant is using to determine if the table exists or not

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: I don't agree, normally we could use the table prefix configured on yesSql side, it should have already use the same configured separator on OC side (or its default value `_`).

@hishamco for your question I saw somewhere, yesSql IConfiguration is not registered in the DI.

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: In summary as I can see ATM, `YesSqlOptions` should be only used to configure the store, then we can always retrieve values from the store.Configuration, unless there is no store registered yet, as we use for the tenant validator.

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: Oops, and tenant validator uses `IDbConnectionValidator` so looks like here the `IStore` can't be injected, and that all need to retrived from `YesSqlOptions` that may need to be completed ;)

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: @jtkech I don't think the prefix table separator is a configurable value in YesSql. AFAIK, you can't retrieve it from YesSql. I think you can get the prefix from YesSql which is not tableSeperator. TableSeperator is something configurable is OC and could be different from one tenant to another. However, in the DbConnectionValidator we assume that the connection being validated will use the currently configured table separator.

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: > I don't think the prefix table separator is a configurable value in YesSql

Yes, I was just meaning that the yesSql store config TablePrefix includes it, but if here we are in the Default tenant to validate the connection of a new tenant, we have to use the tablePrefix param passed to the `ValidateAsync()` and then yes the separator from the YesSql options.

But here we will get the table separator from the YesSqlOptions related to the Default tenant, that's okay if we assume that we only configure these options for the Default tenant, and because we don't use this separator as a tenant settings in the setup form / settings editor.

> TableSeperator is something configurable is OC and could be different from one tenant to another.

Oops ;)

> However, in the DbConnectionValidator we assume that the connection being validated will use the currently configured table separator.

Yes, but you see the little issue.

Just saw that there is a new option field `UseLegacyIdentityColumn`, same remarks, e.g. if it is selectable in the new setup screens, tenant settings editor, okay we will be able to pass it as a new param to `ValidateAsync()`, and in that case not part of the YesSqlOptions (or yes just to show/use a default). Note:  For the separator, one solution would be to do the same. Otherwise here same little issue.

Hmm, maybe here the right solution is to use the dotnet `IConfiguration` to retrieve the `YesSqlOptions` from the config stack but under the specific section of the new tenant we want to create and of which we know the name.

File: src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs:169]: Ok the, `IdentityColumnType` looks much better

File: src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs:23]: No problem, but I'm not sure why it's named `IdentityColumnSize` not `IdentityColumnType`?

File: test/OrchardCore.Tests/Orchard.Queries/SqlParserTests.cs
Comment (hishamco) [reply test/OrchardCore.Tests/Orchard.Queries/SqlParserTests.cs:35]: > like we check the prefix

Where?!! we are checking the result and the expected SQL

File: src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs
Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: @jtkech for your [comment](https://github.com/OrchardCMS/OrchardCore/pull/12528#discussion_r993979250) we referring to `TablePrefixSeparator` not `TablePrefix`

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: Okay, so if we can have different option values per tenant, and because here we may validate the connection of a tenant that doesn't exist yet, we need to find a way to retrieve the possible configured values for this tenant.

ShellSettings holds tenant settings and a tenant config (the injectable ShellConfiguration), for a new tenant the config values are pre-populated (here only internally) but not yet from an eventual tenant specific section.

Hmm, it may become a little more complex, if we use `IConfiguration` by specifying an eventual tenant specific section, if this section doesn't exist it will not inherit of the values of the parent section.

So looks like we need to use `IShellSettingsManager.LoadSettingsAsync(string tenant)`, look at the code to see how we stack the config sources. Note: Needs to be tried and see if it doesn't fail.

Note: Hmm, for sqlite options make sense that they are only global options, about using connexion pool or not.

Finally, for options that will be part of the setup form / tenant editor, better to manage them through the tenant settings, the ones we can retrieve by using shellSettings[""] from the injectable ShellSettings. They use the same kind of config sources as pre-configured values e.g. to pre-populate the setup form / tenant editor, and on saving updated in the tenant specific appsettings that will always win.

So in that case we would not retrieve these values from YesSqlOptions but from the provided form data (or any other binding e.g. for Web Api), as I think we already do for the pre-configurable tenant TablePrefix. And then when configuring the store we would need to use shellSettings[""] to retrieve the actual values, as for TablePrefix here also.

Sorry for the long comments, not so obvious as I need to retrieve myself how it works. Yes would be less complex if we could assume that these yesSql options are the same for all tenants, if not no luck ;)

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: @jtkech @hishamco i think for now so we don’t block this PR, we should just worry about adding this new setting to OC. We can create other PR for later version on OC along with support for Schema. I suggest we only get the new version on YesSql working.
I think the DbConnectionValidator should be refactored one more time to accept ValidationRequestOptions where the consumer would provide the connection string along with other parameters like TablePrefix and TablePrefixSeperator and TableNameConvention.

Comment (jtkech) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: No problem to find a quick compromise to not block this PR.

Then we will have to confirm if YesSqlOptions can be per tenant or not, in both case will be quite easy to do now that we better understand/remember what we already have, this without adding additional options classes.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: @sebastienros what's the action needed here to merge this PR for `1.5` shall we inject both interfaces for now, then refactor this class later on, or shall we need to do other thing

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql/DbConnectionValidator.cs:36]: I will pass a `null` for the schema now, and leave everything as it was

File: src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs:23]: @hishamco 

I think we should remove this from this PR. This can cause a long debate and should not be blocking to this PR.

When this option is added, the value would have to be stored in the `ShellSettings` for each tenant specially when the value is changed to `Int64`. otherwise tenants with long values in their databases could have overflow exception if the options changed again.

IdentitySize + schema option would be part of a separate PR after releasing 1.5.

Comment (hishamco) [reply src/OrchardCore/OrchardCore.Data.YesSql.Abstractions/YesSqlOptions.cs:23]: Furthermore I think the name `IdentityColumnSize` is not suited

**Review by MikeAlhayek — APPROVED** [review]

**Comment by jtkech** [comment]
Okay, I will tweak the DbConnectionValidator and suggest a PR this week end.

@sebastienros just for confirmation should we support YesSqlOptions (e.g. table separator) per tenant, I assume yes, but if it is not worth to do, having global options for all tenants would be simpler.

This because DbConnectionValidator is also used to validate the connection of a tenant that doesn't exist yet, anyway in both cases not so hard to do but just to confirm if it is worth to support.

**Comment by hishamco** [comment]
I might say that we need to support `YesSqlOptions` as global level, but if we could override it per tenant - if it's required - will be a big plus, because in some cases we might need to use another options for a particular tenant(s)

**Comment by jtkech** [comment]
So okay, as a reminder

Any value that will be selectable in a form (or other data source), as I think the `Schema` string will be, will have to be managed through model <=> tenant settings as we do for `TablePrefix`, and that we will pass to the `IDbConnectionValidator.ValidateAsync()`, if too many params maybe pass the `ShellSettings`.

- So will need to update the setup form, tenant editor, tenant API and so on.

The other options being in `YesSqlOptions` as the table separator, and as I think the new Id column size will be, and then used to configure the yesSql store. Here the problem is that

- If we still want to support different values per tenant and if we still want to validate the connection before a tenant is build, I had no solution ;)

- Hmm maybe I understand the problem, I think we should not allow a module that may be enabled or not to collaborate to the `YesSqlOptions` used to work with the database.

-  So looks like `YesSqlOptions` should come from configuration only as we assume for `SqliteOptions` which we bind oob from configuration, configured when calling `AddDataAccess()`. So I think we need to do the same for `YesSqlOptions`, in place of doing a post configure to let other modules collaborate.

- Then the solution to validate the connection before a tenant is built will be to check if a tenant configuration section exists for a given tenant name and so on until fallback to the parent section.

**Pull request merged**
